timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/common/persistence/RedisPersistenceHandler.java

```diff
@@ -0,0 +1,198 @@
+        return overflowedEvents;
+    }
+
+    @Override
+    public Future<?> persistBulkRequest(DbBulkRequest dbBulkRequest, int bulkNum) {
+        return executorService.submit(() -> {
+            LOG.info("Bulk #{} Pushing bulk request to Redis for the {}th time.", bulkNum, dbBulkRequest.getTimesFetched() + 1);
+            Map<String, DbBulkRequest> map = new HashMap<>();
+            String key = FAILED_BULK_PREFIX + UUID.randomUUID().toString();
+            map.put(key, dbBulkRequest);
+            if (!(redisService.pushToRedisList(FAILED_BULKS_QUEUE_NAME, key) && redisService.pushToRedisHash(map))) {
+                LOG.error("Bulk #{} Failed to persist bulk request to Redis", bulkNum);
+            } else {
+                LOG.info("Bulk #{} Bulk request was pushed successfully to Redis.", bulkNum);
+            }
+        });
+    }
+
+    @Override
+    public void persistEvents(ArrayList<Event> events) {
+        Map<String, ArrayList<Event>> map = new HashMap<>();
+        String key = OVERFLOW_EVENTS_PREFIX + UUID.randomUUID().toString();
+        map.put(key, events);
+        if (!(redisService.pushToRedisList(OVERFLOWED_EVENTS_QUEUE_NAME, key) && redisService.pushToRedisHash(map))) {
+            LOG.error("Failed to persist overflowed events list to Redis");
+        } else {
+            LOG.info("List of {} overflowed events was pushed successfully to Redis.", events.size());
+        }
+    }
+
+    @Override
+    public boolean hasFailedBulks() {
+        int iteration = 0;
+        int iterationSize = 100;
+
+        List<String> ids = new ArrayList<>();
+        Map<String, DbBulkRequest> failedBulkRequests = new HashMap<>();
+        do {
+            ids.clear();
+
+            ids.addAll(redisService.getRangeFromRedisList(FAILED_BULKS_QUEUE_NAME, iterationSize * iteration, iterationSize * (iteration + 1) - 1));
+            failedBulkRequests.putAll(redisService.getFromRedis(ids));
+            iteration += 1;
+        }
+        while (areAllIdsExpired(ids, failedBulkRequests));
+
+        return failedBulkRequests.size() > 0;
+    }
+
+    @Override
+    public boolean isCreatedSuccessfully() {
+        boolean connected = redisService.isConnected();
+        if (!connected){
+            LOG.error("Redis wasn't initialized successfully.");
+        }
+        return connected;
+    }
+
+    @Override
+    public long failedBulksAmount() {
+        // including expired failed bulks
+        return redisService.getListLength(FAILED_BULKS_QUEUE_NAME);
+    }
+
+    @Override
+    public long overFlowedEventsListsAmount() {
+        // including expired overflowed events
+        return redisService.getListLength(OVERFLOWED_EVENTS_QUEUE_NAME);
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public void reset() {
+        List<String> queues = Arrays.asList(OVERFLOWED_EVENTS_QUEUE_NAME, FAILED_BULKS_QUEUE_NAME);
+        redisService.deleteFromRedis(queues);
```

Comment:
in this way, the mechanism won't have access to the existing bulks keys
so at some point they will just be expired